### ISSUE TITLE
Provider should have a validating webhook

### ISSUE THREAD

**Issue by darkmuggle** [crossplane-contrib/provider-upjet-aws#963]
### What happened?

provider-aws-ec2 silently ignores invalid configuration. If a user has automated CI/CD like ArgoCD and is creating a managed resource, the resource will be stuck in a never-ending reconciliation loop (ArgoCD patches the object, provider silently ignores the patch ....). 

The reasons that this is a bug is that any automated deployment, or even a Crossplane composition, that relies on the MR to have expected values will be unable to reconcile the object. In the case of ArgoCD a single MR in this situation caused 510 API calls (GET, PATCH, GET, GET, <repeat>) in a 5 minute period. 

### How can we reproduce it?

```
$ kubectl apply -f -<<EOM
---
apiVersion: ec2.aws.upbound.io/v1beta1
kind: Instance
metadata:
  labels:
    horcrux.upbound.io/name: argocd
  name: vm-horcrux-bad-1
spec:
  forProvider:
    ami: ami-0c2644caf041bb6de
    associatePublicIpAddress: true
    instanceType: t2.micro
    region: us-west-2
    keyname: "keyname-should-be-keyName"
EOM

$ kubectl get instance.ec2.aws.upbound.io/vm-horcrux-bad-1 -o yaml | grep keyname

{"apiVersion":"ec2.aws.upbound.io/v1beta1","kind":"Instance","metadata":{"annotations":{},"labels":{"argocd.argoproj.io/instance":"upbound-bh-horcrux","horcrux.upbound.io/name":"argocd"},"name":"vm-horcrux-bad-1"},"spec":{"forProvider":{"ami":"ami-0c2644caf041bb6de","associatePublicIpAddress":true,"instanceType":"t2.micro","keyname":"keyname-should-be-keyName","region":"us-west-2"}}}
```

Notice that `keyname` is not returned, however, the original value is shown in ` kubectl.kubernetes.io/last-applied-configuration` annotation.

### What environment did it happen in?

* Crossplane Version: v1.13.2-up.2
* Provider Version: xpkg.upbound.io/upbound/provider-aws-ec2:v0.43.1
* Kubernetes Version: v1.26.10
* Kubernetes Distribution: Vanilla Kubernetes

**Comment by jbw976** [comment]
perhaps a naive question, but is there a general approach in the api-server that can be done here instead of having to implement/deploy a validating webhook for each provider? e.g. something that tells the api-server to reject resources that have unknown fields?

can't tell quickly if that's possible or not with https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#field-pruning 🤔

**Comment by darkmuggle** [comment]
@jbw976 tbh, a validating webhook request was _probably presumptuous_ on my part. Pruning would certainly work in the ArgoCD/Flux use-cases. My reasoning for requesting a validating was to ensure the correctness of the request.

**Comment by jeanduplessis** [comment]
Here's a summary of an internal conversation on the topic today in the team:

Jared's suggestion seems to make the most sense here. The API server already has the capability to reject REST resources that do not conform to the designated schema (in this case the CRD’s OpenAPIv3 schema). When trying the supplied erroneous manifest, you get the expected 400 - Bad Request error from the API server (kubectl by default uses the strict mode of validation):

```
❯ kubectl apply -f -<<EOM
---
apiVersion: ec2.aws.upbound.io/v1beta1
kind: Instance
metadata:
  labels:
    horcrux.upbound.io/name: argocd
  name: vm-horcrux-bad-1
spec:
  forProvider:
    ami: ami-0c2644caf041bb6de
    associatePublicIpAddress: true
    instanceType: t2.micro
    region: us-west-2
    keyname: "keyname-should-be-keyName"
EOM

Error from server (BadRequest): error when creating "STDIN": Instance in version "v1beta1" cannot be handled as a Instance: strict decoding error: unknown field "spec.forProvider.keyname"
```

You can disable the schema validation when using kubectl with the --validate cmd-line option but the default is strict. 

Could it be that when ArgoCD is applying the (REST) resource configuration (i.e., when applying the Instance.ec2 manifest) in this specific case, the field validation is being skipped? 

Field validation can be performed at the client side (e.g., kubectl fetches the schema for the CRD and validates its input against it), or ServerSideFieldValidation is also a thing, i.e., the API server validates the supplied input. 

ServerSideFieldValidation is at beta level in Kubernetes 1.25 & 1.26 and if the client asks for schema validation, it can be performed at the API server level (the ArgoCD client might not have implemented client-side validation but the API server in use has the ServerSideFieldValidation feature enabled by default and is capable of validating the supplied configuration fields). 

A guess then is that ArgoCD (API server) client could just be explicitly disabling schema validation by passing a &fieldValidation=Ignore query parameter when doing the request.

TL;DR: we should be able to get field validation at the API server (as Jared mentioned) without a validating webhook.

**Comment by yebolenko** [comment]
any update on this?
do you expect that argocd also has to do some changes?
